File: android/interactable/src/main/java/com/wix/interactable/InteractablePoint.java
package com.wix.interactable;

import android.graphics.Point;
import android.graphics.PointF;
import android.util.Log;

import java.util.ArrayList;

/**
 * Created by zachik on 12/02/2017.
 */

public class InteractablePoint {
    public String id;
    public float x;
    public float y;
    public float damping;
    public float tension;
    public float strength;
    public float falloff;
    public InteractableArea influenceArea;

    public InteractablePoint(String id, float x, float y, float damping, float tension, float strength, float falloff, InteractableArea influenceArea) {
        this.id = id;
        this.x = x;
        this.y = y;
        this.damping = damping;
        this.tension = tension;
        this.influenceArea = influenceArea;
        this.strength = strength;
        this.falloff = falloff;
    }

    public PointF positionWithOrigin() {
        PointF res = new PointF(0,0);
        if (this.x != Float.MAX_VALUE) res.x += this.x;
        if (this.y != Float.MAX_VALUE) res.y += this.y;
        return res;
    }

    public float distanceFromPoint(PointF point) {
        float dx = Float.MAX_VALUE;
        float dy = Float.MAX_VALUE;
        if (this.x != Float.MAX_VALUE) dx = point.x -  this.x;
        if (this.y != Float.MAX_VALUE) dy = point.y - this.y;
        if (dx == Float.MAX_VALUE && dy == Float.MAX_VALUE) return Float.MAX_VALUE;
        if (dx == Float.MAX_VALUE) return Math.abs(dy);
        if (dy == Float.MAX_VALUE) return Math.abs(dx);
        return (float) Math.sqrt(dx*dx + dy*dy);
    }

    public static Point deltaBetweenPointAndOrigin(Point point,Point origin) {
        return new Point(point.x - origin.x, point.y - origin.y);
    }

    public static InteractablePoint findClosestPoint(ArrayList<InteractablePoint> pointArray,
                                                     PointF relativePoint) {
//        Log.d("InteractableView","findClosestPoint size = " + pointArray.size());
        float minDist = Float.MAX_VALUE;
        InteractablePoint closestPoint = null;
        for (InteractablePoint point : pointArray) {
//            Log.d("InteractableView","findClosestPoint point x = " + point.x);

            float curDist = point.distanceFromPoint(relativePoint);
            if (curDist < minDist) {
                minDist = curDist;
                closestPoint = point;
            }
        }
        return closestPoint;

    }


}

File: lib/src/InteractablePoint.ts
import { InteractableArea } from "./InteractableArea";
import { IPoint } from "./types";

const MAX = Infinity;

export class InteractablePoint {
    x: number;
    y: number;
    damping: number;
    tension: number;
    strength: number;
    falloff: number;
    id: string;
    influenceArea: InteractableArea;
    haptics: boolean;

    positionWithOrigin(origin: IPoint): IPoint {
        const result = origin;
        if (this.x != Number.MAX_SAFE_INTEGER) result.x += this.x;
        if (this.y != Number.MAX_SAFE_INTEGER) result.y += this.y;
        return result;
    }

    distanceFromPoint(point: IPoint, origin: IPoint): number {
        let dx = Number.MAX_SAFE_INTEGER;
        let dy = Number.MAX_SAFE_INTEGER;
        if (this.x != Number.MAX_SAFE_INTEGER) dx = point.x - (origin.x + this.x);
        if (this.y != Number.MAX_SAFE_INTEGER) dy = point.y - (origin.y + this.y);
        if (dx === Number.MAX_SAFE_INTEGER && dy === Number.MAX_SAFE_INTEGER) return Number.MAX_SAFE_INTEGER;
        if (dx === Number.MAX_SAFE_INTEGER) return Math.abs(dy);
        if (dy === Number.MAX_SAFE_INTEGER) return Math.abs(dx);
        return Math.sqrt(dx*dx + dy*dy);
    }

    static deltaBetweenPoint(point: IPoint, origin: IPoint): IPoint {
        return { x: point.x - origin.x, y: point.y - origin.y };
    }

    static findClosestPoint(points: InteractablePoint[], relativeToPoint: IPoint, origin: IPoint): InteractablePoint {
        let resultPoint = null;
        let minDistance = Number.MAX_SAFE_INTEGER;
        for (const point of points) {
            const distance = point.distanceFromPoint(relativeToPoint, origin);
            if (distance < minDistance) {
                minDistance = distance;
                resultPoint = point;
            }
        }

        return resultPoint;
    }
}